feat(bindings): add SQS AWS Bindings

dpwdec opened the pull request:
## About the PR
Add SQS bindings

### Changelog
* Added SQS Bindings according to https://github.com/asyncapi/bindings/pull/84
* Added `StringOrStringList` type

### Related Issues
None

dpwdec commented on src/LEGO.AsyncAPI.Bindings/Sqs/RedrivePolicy.cs:
```diff
@@ -0,0 +1,32 @@
+
+namespace LEGO.AsyncAPI.Bindings.Sqs
+{
+    public class RedrivePolicy : IAsyncApiElement
+    {
+        /// <summary>
+        /// Prevent poison pill messages by moving un-processable messages to an SQS dead letter queue.
+        /// </summary>
+        public Identifier DeadLetterQueue { get; set; }
+
+        /// <summary>
+        /// The number of times a message is delivered to the source queue before being moved to the dead-letter queue.
+        /// </summary>
+        public int? MaxReceiveCount { get; set; }
+
+        public void Serialize(IAsyncApiWriter writer)
+        {
+            if (writer is null)
+            {
+                throw new ArgumentNullException(nameof(writer));
+            }
+
+            writer.WriteStartObject();
+            writer.WriteRequiredObject("deadLetterQueue", this.DeadLetterQueue, (w, q) => q.Serialize(w));
+            writer.WriteOptionalProperty("maxReceiveCount", this.MaxReceiveCount);
+            writer.WriteEndObject();
+        }
+    }
+}
```
@VisualBean `redrivePolicy` is shared between SNS/SQS binding definitions and has exactly the same properties. Same for `policy` and `statement`. Could these be shared in an AWS folder?

VisualBean replied:
Personally I would copy it. 
No need to code-reuse unless they 'really' depend on each other. They should be singular units of work, so to speak.

VisualBean commented on src/LEGO.AsyncAPI.Bindings/Sqs/SqsChannelBinding.cs:
```diff
@@ -0,0 +1,42 @@
+
+        protected override FixedFieldMap<SqsChannelBinding> FixedFieldMap => new()
+        {
+            
```
missing map

VisualBean commented on src/LEGO.AsyncAPI.Bindings/Sqs/Statement.cs:
```diff
@@ -0,0 +1,77 @@
+        Deny,
+    }
+
+    public class StringOrStringList : IAsyncApiElement
```
this string list could be moved out though. just up a level.
I think there are a couple of bindings with similar semantics.

dpwdec commented on src/LEGO.AsyncAPI.Bindings/Sqs/SqsChannelBinding.cs:
```diff
@@ -0,0 +1,96 @@
+            { "deadLetterQueue", (a, n) => { a.DeadLetterQueue = n.ParseMap(this.queueFixedFields); } },
+        };
+
+        private FixedFieldMap<Queue> queueFixedFields => new()
+        {
+            { "name", (a, n) => { a.Name = n.GetScalarValue(); } },
+            { "fifoQueue", (a, n) => { a.FifoQueue = n.GetBooleanValue(); } },
+            { "deliveryDelay", (a, n) => { a.DeliveryDelay = n.GetIntegerValue(); } },
+            { "visibilityTimeout", (a, n) => { a.VisibilityTimeout = n.GetIntegerValue(); } },
+            { "receiveMessageWaitTime", (a, n) => { a.ReceiveMessageWaitTime = n.GetIntegerValue(); } },
+            { "messageRetentionPeriod", (a, n) => { a.MessageRetentionPeriod = n.GetIntegerValue(); } },
+            { "redrivePolicy", (a, n) => { a.RedrivePolicy = n.ParseMap(this.redrivePolicyFixedFields); } },
+            { "policy", (a, n) => { a.Policy = n.ParseMap(this.policyFixedFields); } },
+            { "tags", (a, n) => { a.Tags = n.CreateSimpleMap(s => s.GetScalarValue()); } },
+        };
```
@VisualBean So, since I added `Extensions` to the `Queue` object I'm getting an `Value cannot be null.` exception from the `ParseField` function. I'm pretty sure this is because of a problem with `x-` not being represented in the `FixedField` map, or it seems like the `ParseMap` is expecting some additional property to exist... I'm not sure how to solve it though as I don't see extension reading examples in the other binding tests.

dpwdec replied:
I think it might actually be overkill to add extensions to these sub properties as well. But I'm still curious if its possible and why it didn't work for me.

VisualBean replied:
The extension should be read it, automatically in the bindingparser (it's another static fieldmap that only contains "x-" and ensure extensions get read. 

I'll have a look a bit later today. See if i can't find a fix for that, and the 'Any' thing we discussed.

dpwdec replied:
Thanks! :)

VisualBean replied:
Ooh so for the extension thing, you want to use extensions further down than the 'root' of the binding?

VisualBean replied:
sorry I didn't understand at first. 
So you want extensions outside of the root of the binding.
It doesn't work as the LoadExtensions method is internal.
<img width="627" alt="image" src="https://github.com/LEGO/AsyncAPI.NET/assets/5294032/0d9bc233-87dd-420c-b855-23eaf899c581">

dpwdec replied:
@VisualBean Hi Alex, yeah I've been looking at the specifications and they don't explicitly have `patternProperties` added, but I think this is not intentional, rather it wasn't considered that `x-` properties might be needed. But, for example, say you have the operation binding for `sqs` like so
```yaml
sqs
  queues:
    - name: "my-queue"
      type: standard
      maxReceiveCount: 2
   - name: "my-queue_error"
      maxReceiveCount: 10
```

Now, if I want to say, specify an extension for the `_error` queue above, it becomes very awkward to do without a per object `x-` extension, it becomes something like:
```yaml
sqs
  queues:
    - name: "my-queue"
      type: standard
      maxReceiveCount: 2
   - name: "my-queue_error"
      maxReceiveCount: 10
  x-someAdditionalProperty:
    queues:
      - key: no-op # don't want to add property to first queue element so need some no-op value
      - dlqParsingKeyAddition: "additive"
```

Instead of just, a much less verbose:
```yaml
sqs
  queues:
    - name: "my-queue"
      type: standard
      maxReceiveCount: 2
   - name: "my-queue_error"
      maxReceiveCount: 10
      x-dlqParsingKeyAddition: "additive"
```

This kind of holds true for a lot of configuration objects in bindings, where its more intuitive for users to use extensions directly on the objects. Does this make sense? I think I will explicitly update the SNS/SQS bindings specs to allow for this extensibility. Is it possible to support this type of behaviour? as I do think it's going to be needed by more people going forward.

VisualBean replied:
We can always add the support. 
Lets hear the binding community what they think about the extensions (it sounds kind of like what the attributes thing was for)

dpwdec replied:
Cool. Presumably if there is approval from the AsyncAPI binding maintainers / community on binding specifications with `patternProperties` added to the child objects in the specification then support for it could move forward. As right now, its explicitly defined as *not* offering that extensibility. Which I believe is an oversight. But I will go back and poll the maintainers and architects there to see if there was an intentional reason that it was held off.

dpwdec replied:
The `attributes` thing I'm not sure about. I think that's just a property that's mirroring the AWS API for filter policy? Rather than an formal extension to the specification 🤔

VisualBean replied:
> Cool. Presumably if there is approval from the AsyncAPI binding maintainers / community on binding specifications with `patternProperties` added to the child objects in the specification then support for it could move forward in AsyncAPI.NET as it would part of the spec. As right now, it's explicitly defined as *not* offering that extensibility. Which I believe is an oversight. But I will go back and poll the maintainers and architects there to see if there was an intentional reason that it was held off.

To be honest, i don't anyone expected bindings to be nested objects. Also, the spec doesn't really care it only denotes bindings as a group, not the individual components. 

However, i think it makes sense to have extension support in all levels -
Effectively; any object should have extension support. I atleast can't see a reason why not to have it.

I am curious about what the maintainers think though.

dpwdec replied:
> I am curious about what the maintainers think though.

Having spoken to them they seem pretty ambivalent, they want to ask the maintainer of those bindings, which is essentially me and maybe Ian Cooper as well. I've made the case to them there about why heavily nested bindings would benefit from individual object extensibility and they seem happy with it so I've updated the nested objects to allow for extensions. It's all on Coops' branch as well so he will have oversight on it and see the changes as he wrote the original binding proposal.

I will also raise a PR on the kafka binding and the ask the kafka binding maintainers if they have an opinion for `topicConfiguration`, my feeling would be though, as it is not heavily nested or listified, they may not see the need, but I'll check anyway.

In any case, I think this functionality is needed to support the binding definitions for SNS/SQS and potentially other nested bindings in future. Let me know if you need me to do anything else to facilitate it in AsyncAPI.NET or if you are happy to move forward with that change yourself. I know I'm probably creating a lot of work for you 😅 so I'm happy to do anything I can to help if its too much.

I think once that is done we should be almost ready to merge 🥳

VisualBean replied:
I'll see if I can't extend the custombindings example with sub extensions.

VisualBean replied:
I'll see if I can't extend the custombindings example with sub extensions.

dpwdec commented:
Hi @VisualBean, Sorry for posting on this PR, it seemed a bit formal to create an issue but I can convert this comment into that if required. 

I wanted to move onto start using the custom binding functionality on some internal bindings, but it seems that the [LEGO.AsyncAPI.Bindings](https://github.com/LEGO/AsyncAPI.NET/tree/main/src/LEGO.AsyncAPI.Bindings) package hasn't been released to Nuget yet. And the main package hasn't been released in a while either. Do you have anything against doing a release? As it seems like there have been substantial changes since the last release, or is there something specific you're waiting for? (Totally fine if so, just want to be aware of the time line).

VisualBean commented:
I wanted your new bindings in, before making a 4.0 release. 

I have a working example of the extensions thing we talked about, so it's definitely possible. 

I am currently house-renovating why my answers and code pushes hasbeen sparse. 

I'll push my extension changes this evening and we can get that change into your PR's and THEN i can push a V4 for you

dpwdec commented:
@VisualBean Thanks for the update. Absolutely no pressure dude! I’m not blocked. For the moment I’m just referencing a local version of the library until the new Nuget package is ready. Good luck with the house renovations! 🙏

VisualBean commented:
> @VisualBean Thanks for the update. Absolutely no pressure dude! I’m not blocked. For the moment I’m just referencing a local version of the library until the new Nuget package is ready. Good luck with the house renovations! 🙏

Finally had a bit of time.
Ive pushed a new "CustomBinding" change, that shows how to utilise extensions. 
This only grants us 1 additional level of extensions, but it's the best I can do with the very sporadic time I have, until late next week. Sorry for the hold up

dpwdec commented on src/LEGO.AsyncAPI.Bindings/Sqs/Queue.cs:
```diff
@@ -0,0 +1,116 @@
+            var queue = new Queue();
+            foreach (var property in mapNode)
+            {
+                property.ParseField(queue, fixedFieldMap, ExtensionHelpers.GetExtensionsFieldMap<Queue>());
```
@VisualBean Thanks for this! Been away but getting back into it now. Tried out your example but I'm getting an error here that `GetExtensionsFieldMap` is internal, it seems to work from within the test project but not in the separate bindings package. Does it need to be public?

Also, when you say

> 1 additional level of extensions

do you mean

- extensions within extensions are unsupported? or,
- extensions of an AsyncApiExtensible object within an AsyncApiExtensible object are unsupported?

VisualBean replied:
> extensions of an AsyncApiExtensible object within an AsyncApiExtensible object are unsupported?
this - atleast in the current 'simple' way of doing it.

Im creating an issue, to ensure all AsyncApiExtensible are automatically populated with any extensions, so we dont need to explicitly do it, in the readers.

VisualBean replied:
Ive made pre-releases for v4, so you can start using them. - will wait with the official release till these bindings are in.

dpwdec replied:
> Not being exhausted from renovating helps.

🥺 well I hope you have a beautiful home to live in now, thank you for still updating stuff while it's been happening.

I'll do some more today and see if I can match it up with the spec.

dpwdec commented:
Hi @VisualBean, I think this PR is now matching the state described in the AsyncAPI bindings repo, so I'm happy to proceed to merge. Let me know if there's any further clean up etc. Also, I'm not sure how to fill out the changelog and issues section so would appreciate some guidance here.

Once it is merged I will move forward with SNS.

VisualBean commented on src/LEGO.AsyncAPI.Bindings/Sqs/SqsChannelBinding.cs:
```diff
@@ -0,0 +1,81 @@
+
+        private FixedFieldMap<Queue> queueFixedFields => new()
+        {
+            { "name", (a, n) => { a.Name = n.GetScalarValue(); } },
```
lookjs like $Ref is missing

VisualBean replied:
But I dont quite understand the $Ref in the spec.
> Allows for an external definition of a queue. The referenced structure MUST be in the format of a Queue. If there are conflicts between the referenced definition and this Queue's definition, the behavior is undefined.

Defined where?

VisualBean replied:
Ive raised the same question in the Binding PR

dpwdec replied:
I think this is intended to allow you to define your queue specification as an external file or remotely as an endpoint. So it would still just be AsyncAPI in the same way that message and schema definitions can be `$ref`. For `$ref` to external files do they generally get explicitly defined in the object model like we have on the `Queue` object? I'm not sure how we unit test it.

dpwdec replied:
Thinking about it I am little doubtful on the value of having $ref just for this queue property so I might push for its removal.

VisualBean replied:
We dont have support for external specs currently though.

VisualBean commented on src/LEGO.AsyncAPI.Bindings/Sqs/SqsOperationBinding.cs:
```diff
@@ -0,0 +1,69 @@
+
+    public class SqsOperationBinding : OperationBinding<SqsOperationBinding>
+    {
+        public List<Queue> Queues { get; set; }
```
missing documentation
> Queue objects that are either the endpoint for an SNS Operation Binding Object, or the deadLetterQueue of the SQS Operation Binding Object

VisualBean commented on src/LEGO.AsyncAPI.Bindings/StringOrStringList.cs:
```diff
@@ -0,0 +1,48 @@
+
+namespace LEGO.AsyncAPI.Bindings
+{
+    public class StringOrStringList : IAsyncApiElement
```
I would really appreciate a few unit tests for this

dpwdec commented:
@VisualBean I think I've addressed your above comments now.

- `DollarRef` removed
- Docs added
- tests for `StringOrStringList`

dpwdec commented:
@VisualBean tests should be fixed now, I think integrating master I messed up some merges.

VisualBean commented:
Great work @dpwdec, thanks for keeping at it!
